nix-mode.el

```diff
@@ -353,27 +373,74 @@ STRING-TYPE type of string based off of Emacs syntax table types"
     (when matching-indentation (indent-line-to matching-indentation) t))
   )
 
-(defun nix-indent-find-BOL-expression-start ()
-  (beginning-of-line)
-  (let ((counter 1))
-    (while (and (> counter 0) (re-search-backward "\\(;\\|=\\|inherit\\|with\\b\\)" nil t))
-      (unless (nix--inside-string-or-comment)
-        (setq counter (cond ((looking-at "with\\|=\\|inherit") (- counter 1))
-                            ((looking-at ";") (+ counter 1))))
-        )
-      )
-      (when (/= counter 0) (goto-char (point-min))) t))
+(defun nix-mode-make-regexp (parts)
+  "Combine the regexps into a single or-delimited regexp."
+  (declare (indent defun))
+  (string-join parts "\\|"))
+
+(defun nix-mode-caps-regexp ()
+  "Return regexp for matching expression caps."
+  (nix-mode-make-regexp nix-mode-caps))
+
+(defun nix-mode-ends-regexp ()
+  "Return regexp for matching expression ends."
+  (nix-mode-make-regexp nix-mode-ends))
+
+(defun nix-mode-quotes-regexp ()
+  "Return regexp for matching string quotes."
+  (nix-mode-make-regexp nix-mode-quotes))
+
+(defun nix-mode-combined-regexp ()
+  "Return combined regexp for matching items of interest."
+    (nix-mode-make-regexp (append nix-mode-caps
+                                  nix-mode-ends
+                                  nix-mode-quotes)))
+
+(defun nix-mode-search-backward ()
+  "Search backward for items of interest regarding indentation."
+  (re-search-backward (nix-mode-combined-regexp) nil t))
 
 (defun nix-indent-expression-start ()
-  (let ((matching-indentation (save-excursion (when (nix-indent-find-BOL-expression-start)
-                                                (current-indentation)))))
-    (when matching-indentation
-      (if (save-excursion (beginning-of-line)
-                          (skip-chars-forward "[:space:]")
-                          (looking-at "let\\|with\\|\\[\\|{\\|("))
-          (indent-line-to matching-indentation)
-        (indent-line-to (+ tab-width matching-indentation)))
-      t)))
+  (let* ((ends 0)
+         (once nil)
+         (done nil)
+         (indent (current-indentation)))
+    (save-excursion
+      ;; we want to indent this line, so we don't care what it contains
+      ;; skip to the beginning so reverse searching doesn't find any matches within
+      (beginning-of-line)
+      ;; search backward until an unbalanced cap is found or no cap or end is found
+      (while (and (not done) (nix-mode-search-backward))
+        (cond
+         ((looking-at (nix-mode-quotes-regexp))
+          ;; skip over strings entirely
+          (re-search-backward nix-mode-quotes-regexp nil t))
```

Comment:
`(re-search-backward (nix-mode-quotes-regexp) nil t))`

Reply:
dang it, lol